### DISCUSSION TITLE
Encoder-based language model in LibriSpeech

### DISCUSSION THREAD

**Discussion by dinoSpeech** [espnet/espnet#5142] [Q&A]
Hello, I have a question for language modeling in LibriSpeech.
I think a language model (LM) is modeling the current word using previous word sequence.
so, for modeling this LM, we use decoder blocks of transformer, because decoder blocks have source attention module and it can prevent LM not to see the future word. 

but in LibriSpeech example, it uses Encoder blocks of transformer for modeling LM. I think encoder-based LM can see all future words, because encoder block doesn't have source attention module.

Why do you guys use encoder for modeling LM? I think decoder-based LM is correct to use for modeling LM in LibriSpeech. 

Best regards

**Comment by sw005320** [comment]
Are you talking about https://github.com/espnet/espnet/blob/master/espnet2/lm/transformer_lm.py#L8?
Yes, we use a transformer encoder, but the important point is to apply a mask in the future subsequences to make a model autoregressive by using https://github.com/espnet/espnet/blob/master/espnet2/lm/transformer_lm.py#L50
So, this model behaves as an autoregressive LM, which is equivalent to your terminology of decoder-based LM without source-target attention.

**Reply by dinoSpeech** [reply]
I check your comment and understand this code now.
thank you for your explain :) 

Best regards
